client/pyscripts/fm11rf08_full.py

```diff
@@ -0,0 +1,942 @@
+# >> "key[][]"
+# >> mad!
+#==============================================================================
+def verifyKeys():
+	global  key
+	global  mad
+
+	badk = 0
+	mad  = False
+
+	lprint(f"{prompt} Check keys..")
+
+	for sec in range (0,16+1):  # 16 normal, 1 dark
+		sn = sec
+		if (sn > 15):  sn = sn + 16
+
+		for ab in [0, 1]:
+			bn  = (sec * 4) + 3
+			if bn >= 64:  bn += 64
+
+			cmd = f"hf mf rdbl -c {ab} --key {key[sec][ab].hex()} --blk {bn}"
+			lprint(f"{prompt}   `{cmd}`", end='', flush=True)
+
+			res = p.console(f"{cmd}")
+			lprint(" " * (3-len(str(bn))), end="")
+			if res == 0:
+				lprint(" ... PASS", end="")
+			else:
+				lprint(" ... FAIL", end="")
+				badk += 1
+				key[sec][ab] = ""
+
+			# check for Mifare Application Directory
+			if (sec == 0) and (ab == 0) \
+			   and (key[0][0] == b'\xa0\xa1\xa2\xa3\xa4\xa5'):
+				mad = True
+				lprint(" - MAD Key")
+			else:
+				lprint("")
+
+	# We need to flush all the output we just collected <shrug>
+	_ = p.grabbed_output
+
+	if badk > 0:
+		lprint(f"{prompt} ! {badk} bad key", end='')
+		lprint("s exist" if badk != 1 else " exists")
+		rv = False
+
+	else:
+		lprint(f"{prompt} All keys verified OK")
+		rv = True
+
+	if mad == True:
+		lprint(f"{prompt} MAD key detected")
+
+	return rv
+
+#+=============================================================================
+# Read all block data - INCLUDING Dark blocks
+# >> blkn
+# >> "data[]"
+'''
+[=]   # | sector 00 / 0x00                                | ascii
+[=] ----+-------------------------------------------------+-----------------
+[=]   0 | 5C B4 9C A6 D2 08 04 00 04 59 92 25 BF 5F 70 90 | \........Y.%._p.
+'''
+#==============================================================================
+def readBlocks():
+	global  data
+	global  blkn
+
+	data = []
+	blkn = list(range(0, 63+1)) + list(range(128, 135+1))
+
+	# The user   uses keyhole #1 (-a)
+	# The vendor uses keyhole #2 (-b)
+	# The thief  uses keyhole #4 (backdoor)
+	#                   |___
+	rdbl = f"hf mf rdbl -c 4 --key {dkey} --blk"
+
+	lprint(prompt)
+	lprint(prompt + " Load blocks {0..63, 128..135}[64+8=72] from the card")
+
+	bad = 0
+	for n in blkn:
+		cmd = f"{rdbl} {n}"
+		print(f"\r{prompt} `{cmd}`", end='', flush=True)
+
+		for retry in range(5):
+			p.console(f"{cmd}")
+
+			found = False
+			for line in p.grabbed_output.split('\n'):
+				if " | " in line and "# | s" not in line:
+					l = line[4:76]
+					data.append(l)
+					found = True
+			if found:  break
+
+		if not found:
+			data.append(f"{n:3d} | -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- | ----------------")
+			bad += 1
+
+	print(" .. OK")
+
+#+=============================================================================
+# Patch keys in to data
+# >> "key[][]"
+# >> "data[]"
+# >> keyok!
+'''
+  3 | 00 00 00 00 00 00 87 87 87 69 00 00 00 00 00 00 | .........i......
+'''
+#==============================================================================
+def patchKeys(keyok):
```

Comment:
note that dump from recovery is already patched

Reply:
For reasons discussed above, I do not load the existing `dump.bin` so this action is 'required'